File: dotnet/DesignPatterns/CreationalDesignPatterns/FactoryDesignPattern/FactoryDesignPattern.Example1/Factories/NotificationFactory.cs
using FactoryDesignPattern.NotificationAPI.Enums;
using FactoryDesignPattern.NotificationAPI.Services.Abstracts;
using FactoryDesignPattern.NotificationAPI.Services.Concretes;

namespace FactoryDesignPattern.NotificationAPI.Factories;

public class NotificationFactory : INotificationFactory
{
	private readonly IServiceProvider _serviceProvider;

	public NotificationFactory(IServiceProvider serviceProvider)
	{
		_serviceProvider = serviceProvider;
	}

	public INotificationService Create(NotificationType notificationType)
	{
		//return notificationType switch
		//{
		//	NotificationType.Email => new EmailNotificationService(),
		//	NotificationType.Sms => new SmsNotificationService(),
		//	NotificationType.PushNotification => new PushNotificationService(),
		//	_ => throw new ArgumentException("Invalid notification type", nameof(notificationType))
		//};

		return notificationType switch
		{
			NotificationType.Email => _serviceProvider.GetRequiredService<EmailNotificationService>(),
			NotificationType.Sms => _serviceProvider.GetRequiredService<SmsNotificationService>(),
			NotificationType.PushNotification => _serviceProvider.GetRequiredService<PushNotificationService>(),
			_ => throw new ArgumentException("Invalid notification type", nameof(notificationType))
		};
	}
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/FactoryDesignPattern/FactoryDesignPattern.Example1/src/factories/NotificationFactory.ts
import { NotificationType } from '../enums/NotificationType';
import { INotificationFactory } from './INotificationFactory';
import { INotificationService } from '../services/INotificationService';
import { EmailNotificationService } from '../services/EmailNotificationService';
import { SmsNotificationService } from '../services/SmsNotificationService';
import { PushNotificationService } from '../services/PushNotificationService';

export class NotificationFactory implements INotificationFactory {
  create(notificationType: NotificationType): INotificationService {
    switch (notificationType) {
      case NotificationType.Email:
        return new EmailNotificationService();
      case NotificationType.Sms:
        return new SmsNotificationService();
      case NotificationType.PushNotification:
        return new PushNotificationService();
      default:
        throw new Error(`Invalid notification type: ${notificationType}`);
    }
  }
}
